src/UbuntuToolkit/ucubuntushape.cpp

```diff
@@ -1229,7 +1229,7 @@ QSGNode* UCUbuntuShape::updatePaintNode(QSGNode* oldNode, UpdatePaintNodeData* d
         }
         if (provider) {
             QObject::connect(provider, SIGNAL(textureChanged()), this, SLOT(_q_textureChanged()));
-            QObject::connect(provider, SIGNAL(destroyed()), this, SLOT(_q_providerDestroyed()));
+            QObject::connect(provider, SIGNAL(destroyed()), this, SLOT(_q_providerDestroyed()), Qt::DirectConnection);
```

Comment:
This *may* end up in a dead lock and we should never use DirectConnection when it may be distorted from a different thread. 

IMO we should leave this as is, and add _q_providerDestroyed to it's deconstructor that also disconnects all signals before it gets destroyed. You may also add this in a lamda function that check if this is valid, but should not be needed if disconnect is called on deconstruction. Not sure why this has no deconstructor.

Reply:
Signals (both ways) are automatically disconnected on destruction of an object. The problem is that you can't *always* assume that the texture provider will be destroyed when it changes. Instead, it can be removed from this UbuntuShape and given to a different one.

Reply:
`DirectConnection` isn't `BlockingQueuedConnection`. With `DirectConnection`, the slot will be called on whichever thread the signal is emitted. In this case, it's the rendering thread where a texture provider is destructed. To the best of my knowledge, this pointer is accessed only in the rendering thread, thus it's safe to use `DirectConnection` in this case.

For the latter part, I'm not sure which object destruction are you talking about. Could you please clarify?

Reply:
Ah, yes, I see. Ratchanan is correct about this being a `DirectConnection`, not a `BlockingQueuedConnection`. It definitely isn't as safe as the current plain `QueuedConnection`, granted, but it will not cause a deadlock from the signal and slot thread being different.

As for why this can't be done in the texture provider destructor... `destroyed()` is signaled right before the object is destructed. However, destruction occurs in the main thread. That occurs while the render thread is busy in an event loop iteration, so the `destroyed()` signal is not received by the UbuntuShape. So, instead of trying to access `null` as we would if we had acted on `destroyed()`, we try to access the freed pointer that was just a texture provider.

Reply:
>It definitely isn't as safe as the current plain `QueuedConnection`...

No. In this case, it's completely safe, because `updatePaintNode()` is also supposed to run in the rendering thread.

https://doc.qt.io/qt-5/qquickitem.html#updatePaintNode

> Called on the render thread when it is time to sync the state of the item with the scene graph.

I don't understand the latter part. Texture provider is supposed to be destructed in the rendering thread, while `UbuntuShape`'s thread affinity is supposed to be GUI (main) thread, isn't it?

Reply:
Either way, destruction happens before we can act on destroyed.